### PR TITLE
Add kubectl create daemonset command

### PR DESCRIPTION
**Pull request by zjj2wry**
**What this PR does / why we need it**:
https://github.com/kubernetes/kubectl/issues/12
https://github.com/kubernetes/kubernetes/issues/48957

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Add kubectl create daemonset command 
```

### CODE REVIEW

**Comment by NickrenREN** [comment]
> W0816 09:19:55.750] 2017-08-16 09:19:55.749346 I | proto: duplicate proto type registered: google.protobuf.Duration
W0816 09:19:55.750] 2017-08-16 09:19:55.749364 I | proto: duplicate proto type registered: google.protobuf.Timestamp
I0816 09:19:56.059] List of generated docs doesn't match a freshly built list. Please run hack/update-generated-docs.sh
I0816 09:19:56.067] FAILED   hack/make-rules/../../hack/verify-generated-docs.sh	40s

Please run hack/update-generated-docs.sh

**Comment by pwittrock** [comment]
Create `creeate daemonset` command?

**Comment by pwittrock** [comment]
FYI
cc @alexandercampbell

**Comment by zhangxiaoyu-zidif** [comment]
Add kubectl create deamonset command
s/deamonset /daemonset

**Comment by zjj2wry** [comment]
@zhangxiaoyu-zidif thanks

**Comment by zjj2wry** [comment]
cc @kubernetes/sig-cli-pr-approver

**Review by dixudx — CHANGES_REQUESTED** [review]
bump to `apps/v1beta2` instead of `extensions/v1beta1`.

File: pkg/kubectl/daemonset.go
Comment (zjj2wry) [reply pkg/kubectl/daemonset.go]: cool, i will updated this

File: pkg/kubectl/cmd/create_deamonset.go
Comment (zjj2wry) [reply pkg/kubectl/cmd/create_deamonset.go]: yes

File: pkg/kubectl/cmd/create_deamonset.go
Comment (zjj2wry) [reply pkg/kubectl/cmd/create_deamonset.go]: this lower case maybe better

**Comment by zjj2wry** [comment]
updated

File: pkg/kubectl/cmd/util/factory_client_access.go
Comment (dixudx) [reply pkg/kubectl/cmd/util/factory_client_access.go]: @zjj2wry Forgot to update this?

File: pkg/kubectl/daemonset.go
Comment (dixudx) [reply pkg/kubectl/daemonset.go]: @zjj2wry Bump to `apps/v1beta2`

File: pkg/kubectl/daemonset.go
Comment (dixudx) [inline pkg/kubectl/daemonset.go]: Should not use `extensions/v1beta1` now, since we've bumped to `appsv1beta2`.

**Review by dixudx — CHANGES_REQUESTED** [review]

File: pkg/kubectl/daemonset.go
Comment (zjj2wry) [reply pkg/kubectl/daemonset.go]: actually my cluster not support appsv1beta2. i support this appsv1beta2 version, but we should lower version to compatibility？

**Review by apelisse — COMMENTED** [review]
That looks overall good to me. I find it a little dubious that we need to create it for both version though. What do you think @dixudx ?

**Comment by zjj2wry** [comment]
@apelisse will add fallback when high version not exist then use lower version in another pr soon. ptal~

**Comment by dixudx** [comment]
> we need to create it for both version though

@apelisse Yes, we should cover them in tests to ensure the subcommands work as expected as well.

**Comment by zjj2wry** [comment]
i have add some unittest for ```apps/v1beta2``` version. will add some bashtest in `failback` pr ~.~

File: pkg/kubectl/cmd/util/factory_client_access.go
Comment (dixudx) [inline pkg/kubectl/cmd/util/factory_client_access.go]: Change `DaemonsetV1Beta2GeneratorName` to `DaemonsetAppsV1beta2GeneratorName` to make it more explicit?

File: pkg/kubectl/cmd/util/factory_client_access.go
Comment (dixudx) [reply pkg/kubectl/cmd/util/factory_client_access.go]: Change `DaemonsetV1Beta1GeneratorName` to `DaemonsetExtensionsV1Beta1GeneratorName` to make it more explicit?

File: pkg/kubectl/cmd/create_deamonset.go
Comment (dixudx) [inline pkg/kubectl/cmd/create_deamonset.go]: Since we've already bumped to `apps/v1beta2` in apiserver and this is also a fresh new subcommand, I'd suggest using `apps/v1beta2` directly.

File: pkg/kubectl/cmd/util/factory_client_access.go
Comment (zjj2wry) [reply pkg/kubectl/cmd/util/factory_client_access.go]: 👍 agree, will updated. thank you

**Comment by dixudx** [comment]
> I find it a little dubious that we need to create it for both version though.

@apelisse @zjj2wry Since resources in `extensions` are marked deprecated, I think we may not need to support `extensions.v1beta1` in this new subcommand.

**Comment by zjj2wry** [comment]
@dixudx you are right, done. have set "apps/v1beta2" to default. but we can not delete  "extensions/v1beta1". because have mary users use newer version kubectl. but older version cluster. like me~
```
admindeMacBook-Pro-2:kubectl admin$ ./kubectl create daemonset b --image=nginx
WARNING: New deployments generator "apps/v1beta2" specified, but it isn't available. Falling back to "extensions/v1beta1".
daemonset "b" created
```

**Comment by zjj2wry** [comment]
@apelisse ptal

**Comment by apelisse** [comment]
Still looks good to me, what about you @dixudx?

File: pkg/kubectl/cmd/create_deamonset.go
Comment (smarterclayton) [inline pkg/kubectl/cmd/create_deamonset.go]: Please provide a better description of daemonsets here, including what they do.  This should be educational to new users who know nothing about kubernetes.  I.e. try to answer:

1. what is a daemonset
2. what are the key use cases and how will this command help me start one
3. where to go to find more information about daemonsets

These create commands help users bridge the gap from novice to productive.

File: pkg/kubectl/cmd/util/factory_client_access.go
Comment (janetkuo) [inline pkg/kubectl/cmd/util/factory_client_access.go]: batchv2alpha1 does not have "daemonset" resource. appsv1beta2 should be used instead.

File: hack/make-rules/test-cmd-util.sh
Comment (janetkuo) [inline hack/make-rules/test-cmd-util.sh]: Please move this test to `run_daemonset_tests()`, and write `kubectl create daemonset` tests similar to: 
https://github.com/kubernetes/kubernetes/blob/4425841ce2a99dfc20f15d2e4813b635520ef19d/hack/make-rules/test-cmd-util.sh#L2691-L2723

File: pkg/kubectl/cmd/util/factory_client_access.go
Comment (dixudx) [reply pkg/kubectl/cmd/util/factory_client_access.go]: +1. Agree. Since this is falling back to older version, newest version `appsv1beta2` should be used here.

File: pkg/kubectl/cmd/create_deamonset.go
Comment (dixudx) [inline pkg/kubectl/cmd/create_deamonset.go]: Both L88 and L95 use this variable. Move `cmdutil.GetFlagStringSlice(cmd, "image")` outer?

**Comment by zjj2wry** [comment]
done @smarterclayton ptal

File: pkg/kubectl/daemonset.go
Comment (apelisse) [reply pkg/kubectl/daemonset.go]: what version is your cluster? This new code is not going to be released before 1.9, and we don't have to support server below 1.8.

**Comment by zjj2wry** [comment]
ping @janetkuo @smarterclayton for approve

**Comment by zjj2wry** [comment]
/retest
@smarterclayton this PR ready go, ptal

**Comment by zjj2wry** [comment]
ping @smarterclayton ptal

File: pkg/kubectl/daemonset.go
Comment (smarterclayton) [reply pkg/kubectl/daemonset.go]: Yeah, this code should use the new v1 workload api, not extensions.

**Comment by smarterclayton** [comment]
Now that daemonset is going to be v1 in 1.9, we should update this to only use the v1 api.

**Comment by zjj2wry** [comment]
@smarterclayton updated, ptal

**Comment by zjj2wry** [comment]
@smarterclayton this PR ready, ptal

**Comment by smarterclayton** [comment]
Unfortunately this missed 1.9 due to reviewer overload, but this is in a pretty good shape for when 1.10 opens.

**Comment by zjj2wry** [comment]
@smarterclayton not at all, thanks

**Comment by zjj2wry** [comment]
rebased to trigger test

**Comment by zjj2wry** [comment]
ping @smarterclayton @liggitt for approve

**Comment by zjj2wry** [comment]
@kubernetes/sig-cli-pr-reviews

**Comment by zjj2wry** [comment]
@kubernetes/sig-cli-pr-reviews @liggitt i want this can get into v1.10, ptal

**Comment by zjj2wry** [comment]
@kubernetes/sig-cli-maintainers @deads2k @smarterclayton @liggitt PTAL

File: pkg/kubectl/cmd/create_daemonset_test.go
Comment (soltysh) [inline pkg/kubectl/cmd/create_daemonset_test.go]: 2018 these days ;-)

File: pkg/kubectl/cmd/create_deamonset.go
Comment (soltysh) [inline pkg/kubectl/cmd/create_deamonset.go]: 2018

File: pkg/kubectl/cmd/create_deamonset.go
Comment (soltysh) [inline pkg/kubectl/cmd/create_deamonset.go]: I doubt it's necessary to repeat docs here. Link to docs is fine, but I don't recall any create command that has it, so I'd drop it as well. The long description is usually one full sentence what the command does and that's it. The examples is where you focus on usage. All in all if you reach out for `--help` you'll get sufficient information.

File: pkg/kubectl/cmd/create_deamonset.go
Comment (soltysh) [inline pkg/kubectl/cmd/create_deamonset.go]: More examples is definitely good idea.

File: pkg/kubectl/cmd/create_deamonset.go
Comment (soltysh) [reply pkg/kubectl/cmd/create_deamonset.go]: Upper case please, to follow the pattern we have in other create commands. The resource names are always UppperCase'd.

File: pkg/kubectl/cmd/create_deamonset.go
Comment (soltysh) [reply pkg/kubectl/cmd/create_deamonset.go]: Do you support node label, maybe?

File: pkg/kubectl/daemonset.go
Comment (soltysh) [inline pkg/kubectl/daemonset.go]: 2018

File: pkg/kubectl/daemonset.go
Comment (soltysh) [inline pkg/kubectl/daemonset.go]:
```go
if err := b.validate(); err != nil {
    return
}
```

**Review by soltysh — CHANGES_REQUESTED** [review]
A few nits, make sure to addressed them and I'll try to push this forward.

File: pkg/kubectl/cmd/create_deamonset.go
Comment (zjj2wry) [reply pkg/kubectl/cmd/create_deamonset.go]: nothing now，we want to make create command simple enough, and then use other commands to edit it.
```
kubectl create ds hello --image=nginx --dry-run -o yaml | kubectl lable --local -f - env=qa -o yaml | kubectl create -f -
```

**Comment by adohe** [comment]
overall lgtm. @zjj2wry please squash before we can merge, I would like to get this merged before we refactor `create` commands family.

**Comment by zjj2wry** [comment]
@adohe ready go, ptal

**Comment by zjj2wry** [comment]
@adohe @smarterclayton @liggitt rebased, ptal

**Pull request closed without merging**
